File: python/zyvor_janus/benchmarks/twin_store.py
"""Digital twin store for calibrated cluster fingerprints."""

from __future__ import annotations

import json
import sqlite3
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path


@dataclass
class TwinEntry:
    gpu_type: str
    model: str
    ttft_ms: float
    tps: float
    throughput: float
    measured_at: str
    aiperf_run_id: str | None = None


class TwinStore:
    def __init__(self, root: Path) -> None:
        self.root = root
        self.root.mkdir(parents=True, exist_ok=True)
        self.db_path = self.root / "twins.sqlite"
        self._init_db()

    def _init_db(self) -> None:
        with sqlite3.connect(self.db_path) as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS twins (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    gpu_type TEXT NOT NULL,
                    model TEXT NOT NULL,
                    ttft_ms REAL NOT NULL,
                    tps REAL NOT NULL,
                    throughput REAL NOT NULL,
                    measured_at TEXT NOT NULL,
                    aiperf_run_id TEXT,
                    version INTEGER NOT NULL DEFAULT 1
                )
                """
            )

    def upsert(self, entry: TwinEntry) -> int:
        with sqlite3.connect(self.db_path) as conn:
            cur = conn.execute(
                "SELECT MAX(version) FROM twins WHERE gpu_type=? AND model=?",
                (entry.gpu_type, entry.model),
            )
            version = (cur.fetchone()[0] or 0) + 1
            conn.execute(
                """
                INSERT INTO twins (gpu_type, model, ttft_ms, tps, throughput, measured_at, aiperf_run_id, version)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    entry.gpu_type,
                    entry.model,
                    entry.ttft_ms,
                    entry.tps,
                    entry.throughput,
                    entry.measured_at,
                    entry.aiperf_run_id,
                    version,
                ),
            )
            return version

    def latest(self, gpu_type: str, model: str) -> TwinEntry | None:
        with sqlite3.connect(self.db_path) as conn:
            row = conn.execute(
                """
                SELECT gpu_type, model, ttft_ms, tps, throughput, measured_at, aiperf_run_id
                FROM twins
                WHERE gpu_type=? AND model=?
                ORDER BY version DESC LIMIT 1
                """,
                (gpu_type, model),
            ).fetchone()
        if not row:
            return None
        return TwinEntry(*row)

    def detect_drift(self, gpu_type: str, model: str, measured_ttft_ms: float, threshold: float = 0.10) -> bool:
        twin = self.latest(gpu_type, model)
        if twin is None:
            return False
        delta = abs(measured_ttft_ms - twin.ttft_ms) / max(twin.ttft_ms, 1e-6)
        return delta > threshold

    def export_json(self, path: Path) -> None:
        with sqlite3.connect(self.db_path) as conn:
            rows = conn.execute(
                "SELECT gpu_type, model, ttft_ms, tps, throughput, measured_at, aiperf_run_id FROM twins ORDER BY id"
            ).fetchall()
        payload = [
            {
                "gpu_type": r[0],
                "model": r[1],
                "ttft_ms": r[2],
                "tps": r[3],
                "throughput": r[4],
                "measured_at": r[5],
                "aiperf_run_id": r[6],
            }
            for r in rows
        ]
        path.write_text(json.dumps(payload, indent=2))

    @staticmethod
    def now_iso() -> str:
        return datetime.now(timezone.utc).isoformat()

File: crates/zyvor-janus-api/src/twin_store.rs
use std::path::Path;

use rusqlite::Connection;
use serde::Serialize;

/// Read-only port of Python's `TwinStore` (`python/zyvor_janus/benchmarks/twin_store.py`),
/// scoped to what `GET /api/twins` needs. Nothing in this codebase writes twins
/// over HTTP -- the `twins` table is populated by offline AIPerf calibration
/// tooling, so only the read path is ported here.
#[derive(Debug, Clone, Serialize)]
pub struct TwinEntry {
    pub gpu_type: String,
    pub model: String,
    pub ttft_ms: f64,
    pub tps: f64,
    pub throughput: f64,
    pub measured_at: String,
    pub aiperf_run_id: Option<String>,
}

/// Ensures `twins.sqlite` (and its `twins` table) exist at `db_path`, then
/// returns every row ordered by `id`, matching Python's `export_json`.
pub fn list_twins(db_path: &Path) -> rusqlite::Result<Vec<TwinEntry>> {
    if let Some(parent) = db_path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    let conn = Connection::open(db_path)?;
    conn.execute(
        "CREATE TABLE IF NOT EXISTS twins (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            gpu_type TEXT NOT NULL,
            model TEXT NOT NULL,
            ttft_ms REAL NOT NULL,
            tps REAL NOT NULL,
            throughput REAL NOT NULL,
            measured_at TEXT NOT NULL,
            aiperf_run_id TEXT,
            version INTEGER NOT NULL DEFAULT 1
        )",
        [],
    )?;

    let mut stmt = conn.prepare(
        "SELECT gpu_type, model, ttft_ms, tps, throughput, measured_at, aiperf_run_id
         FROM twins ORDER BY id",
    )?;
    let rows = stmt.query_map([], |row| {
        Ok(TwinEntry {
            gpu_type: row.get(0)?,
            model: row.get(1)?,
            ttft_ms: row.get(2)?,
            tps: row.get(3)?,
            throughput: row.get(4)?,
            measured_at: row.get(5)?,
            aiperf_run_id: row.get(6)?,
        })
    })?;
    rows.collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_store_returns_empty_list() {
        let dir = tempfile_dir();
        let db_path = dir.join("twins.sqlite");
        let twins = list_twins(&db_path).unwrap();
        assert!(twins.is_empty());
        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn round_trips_inserted_rows() {
        let dir = tempfile_dir();
        let db_path = dir.join("twins.sqlite");
        list_twins(&db_path).unwrap();

        let conn = Connection::open(&db_path).unwrap();
        conn.execute(
            "INSERT INTO twins (gpu_type, model, ttft_ms, tps, throughput, measured_at, aiperf_run_id, version)
             VALUES ('h100', 'llama-70b', 42.0, 100.0, 5000.0, '2026-01-01T00:00:00Z', 'run-1', 1)",
            [],
        )
        .unwrap();

        let twins = list_twins(&db_path).unwrap();
        assert_eq!(twins.len(), 1);
        assert_eq!(twins[0].gpu_type, "h100");
        assert_eq!(twins[0].aiperf_run_id.as_deref(), Some("run-1"));
        std::fs::remove_dir_all(&dir).ok();
    }

    fn tempfile_dir() -> std::path::PathBuf {
        let dir = std::env::temp_dir().join(format!(
            "zyvor-janus-api-twin-store-test-{}",
            uuid::Uuid::new_v4()
        ));
        std::fs::create_dir_all(&dir).unwrap();
        dir
    }
}
